contracts/gateway/src/contract/execute.rs

```diff
@@ -1,143 +1,175 @@
-        if msgs.is_empty() {
-            return Ok(Response::new());
-        }
-
-        ensure_unique_ids(&msgs)?;
+        let partitioned_msgs = ignore_empty_msgs(msgs)
+            .try_map(check_for_duplicates)?
+            .try_map(|msgs| self.partition_by_verified(msgs))?;
 
-        let (_, unverified) = self.partition_by_verified(msgs)?;
-
-        // short circuit if there are no unverified messages
-        // there is no need to interact with the verifier so it saves gas
-        if unverified.is_empty() {
-            return Ok(Response::new());
+        match partitioned_msgs {
+            None => Ok(Response::new()),
+            Some((verified, unverified)) => self.verify_unverified_messages(verified, unverified),
         }
-
-        Ok(Response::new().add_message(WasmMsg::Execute {
-            contract_addr: self.config.verifier.to_string(),
-            msg: to_binary(&aggregate_verifier::msg::ExecuteMsg::VerifyMessages {
-                messages: unverified,
-            })
-            .change_context(ContractError::CreateVerifierExecuteMsg)?,
-            funds: vec![],
-        }))
     }
 
     pub fn route_messages(
         &mut self,
         sender: Addr,
         msgs: Vec<Message>,
     ) -> Result<Response, ContractError> {
-        if sender == self.config.router {
-            self.route_outgoing_messages(msgs)
+        if sender == self.router.address {
+            Self::route_outgoing_messages(&mut self.store, msgs)
         } else {
             self.route_incoming_messages(msgs)
         }
     }
 
-    fn route_outgoing_messages(&mut self, msgs: Vec<Message>) -> Result<Response, ContractError> {
-        ensure_unique_ids(&msgs)?;
+    fn route_outgoing_messages(
+        store: &mut S,
+        msgs: Vec<Message>,
+    ) -> Result<Response, ContractError> {
+        let msgs = check_for_duplicates(msgs)?;
 
         for msg in msgs.iter() {
-            self.store.save_outgoing_msg(msg.cc_id.clone(), msg)?;
+            store.save_outgoing_msg(msg.cc_id.clone(), msg)?;
         }
 
         Ok(Response::new().add_events(
             msgs.into_iter()
-                .map(|msg| GatewayEvent::MessageRouted { msg }.into()),
+                .map(|msg| GatewayEvent::Routing { msg }.into()),
         ))
     }
 
     fn route_incoming_messages(&self, msgs: Vec<Message>) -> Result<Response, ContractError> {
-        ensure_unique_ids(&msgs)?;
+        let partitioned_msgs = ignore_empty_msgs(msgs)
+            .try_map(check_for_duplicates)?
+            .try_map(|msgs| self.partition_by_verified(msgs))?;
 
-        let (verified, unverified) = self.partition_by_verified(msgs)?;
-
-        let any_verified = !verified.is_empty();
+        match partitioned_msgs {
+            None => Ok(Response::new()),
+            Some((verified, unverified)) => self.route_verified_messages(verified, unverified),
+        }
+    }
 
-        let mut response = Response::new()
+    fn verify_unverified_messages(
```

Comment:
could we extract the events part out from this function or at least rename the function? it seems weird to me that the function name references unverified messages and still accepts both verified and unverified as parameters. Same with `route_verified_messages`

Reply:
I'm open to suggestions for how to do it/alternative names 🙂

Reply:
let me know how you like this modification